### PR TITLE
feat(plugin): add support for zora in collectible

### LINKED ISSUE #4299: [Bounty] Integrate Zora, an NFT auction protocol, into Mask Network
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [ ] Intermediate
  - [x] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks
- Bounty Size
  - [x] 200 $MASK or $2000

## Part 1: Problem Description

**Background**
Zora app is an decentralized auction house 

**Solution**
Create a plugin that integrate Zora.  The basic product logic could resemble our Opensea and Rarible integration. The UI is in a card style with different tabs serving different purposes. 
opensea1
_Opensea Plugin UI_

Default page could be the item overview 
A second tab for auction info
Other tabs for offerings, historical prices etc.

Zora is more artist focused than Opensea and rarible. So the UI design could go along the same line.  DO talk to us before commiting to certain integration design though. Could stick with the opensea layout as well.

Anylink from Zora on Twitter could trigger the plugin ( could have some slightly different treatment). The tab order can be changed, corresponding landing and ending page could be added. Share button for such a social nft platform is a must

Language like 
**I just purchased xxx for yyy dollar, and I feel like on top of the world.** 
Could be added to help spread the word.

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- [Zora doc](https://docs.zora.co/zoraos/)

## Part 3: Contact and Support
@yisiliu for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q
Zora discord server: https://discord.com/invite/Va58aMrcwk

**Submission Requirements**
Please include demo to show the integration is working.

**Winner Announcement Date**
Winners will be announced at the conclusion of GR11

**Judging Criteria**
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by brymut**
## Description
Adding support for Zora in Collectible plugin so can get information about NFTs on Zora.

Closes #4299 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

**Comment by brymut** [comment]
Hi, may I get some help about the build error:
```ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up-to-date with packages/web3-providers/package.json

Note that in CI environments this setting is true by default. If you still need to run install in such cases, use "pnpm install --no-frozen-lockfile"```

File: packages/web3-providers/src/zora/constants.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export const ZORA_MAINNET_GRAPHQL_URL = 'https://indexer-prod-mainnet.zora.co/v1/graphql'
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/constants.ts]: `https://cors.r2d2.to`

File: packages/web3-providers/src/zora/index.ts [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+            name: asset.metadata.json.name ?? '',
+            description: asset.metadata.json.description ?? '',
+            mediaUrl: asset.metadata.json.animation_url ?? asset.metadata.json.image_url ?? '',
+            owner: asset.owner,
+        },
+        tokenId: tokenId,
+    }
+}
+
+export class ZoraApi implements NonFungibleTokenAPI.Provider {
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/index.ts]:
```suggestion
export class ZoraAPI implements NonFungibleTokenAPI.Provider {
```

File: packages/web3-providers/src/zora/index.ts [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    private readonly _apiKey
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/index.ts]: No using?

Comment (brymut) [reply packages/web3-providers/src/zora/index.ts]: `OpenSeaAPI` & `RaribleAPI` both implements

Comment (guanbinrui) [reply packages/web3-providers/src/zora/index.ts]: Only OpenSea has this field cause of the API requirements. But Zora API doesn't have a thing like that. It's  Not necessary.

File: packages/web3-providers/src/zora/index.ts [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+        const orders = nftAsks.Token[0].transferEvents.map((event: ZoraAsk): NonFungibleTokenAPI.AssetOrder => {
+            if (event.transaction.auctionCreatedEvents?.length !== 0) {
+                const ask = first(event.transaction.auctionCreatedEvents)
+                return {
+                    created_time: new Date(`${ask?.blockTimestamp}Z`).getTime().toString(),
+                    approved_on_chain: false,
+                    current_price: formatWeiToEther(ask?.reservePrice || 0).toString(),
+                    payment_token: ask?.auctionCurrency,
+                    payment_token_contract: {
+                        name: '',
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/index.ts]:
```suggestion
                        name: 'Ether',
```

File: packages/web3-providers/src/zora/index.ts [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+
+        const offers = nftBids.Token[0].transferEvents.map((event: ZoraBid): NonFungibleTokenAPI.AssetOrder => {
+            if (event.transaction.marketBidEvents?.length !== 0) {
+                const bid = first(event.transaction.marketBidEvents)
+                return {
+                    created_time: new Date(`${bid?.blockTimestamp}Z`).getTime().toString(),
+                    current_price: bid?.amount,
+                    payment_token: bid?.currencyAddress || '',
+                    payment_token_contract: {
+                        name: '',
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/index.ts]: Does Zora only support ether token?

```suggestion
                        name: 'Ether',
```

File: packages/web3-providers/src/zora/index.ts [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+                    approved_on_chain: false,
+                    maker_account: {
+                        user: { username: bid?.recipient || '' },
+                        address: bid?.recipient || '',
+                        profile_img_url: '',
+                        link: `https://zora.co/${bid?.recipient}`,
+                    },
+                }
+            }
+            return {} as NonFungibleTokenAPI.AssetOrder
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/index.ts]:
```suggestion
            return
```

File: packages/web3-providers/src/zora/index.ts [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+        })
+        return offers.filter((event: NonFungibleTokenAPI.AssetOrder) => {
+            return Object.keys(event).length !== 0
+        })
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/index.ts]:
```suggestion
        return offers.filter(Boolean)
```

File: packages/web3-providers/src/zora/index.ts [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+                            to: {
+                                username: buy?.winner,
+                                address: buy?.winner,
+                                imageUrl: '',
+                                link: `https://zora.co/${buy?.winner}`,
+                            },
+                        },
+                    } as NonFungibleTokenAPI.History
+                }
+                return {} as NonFungibleTokenAPI.History
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/index.ts]:
```suggestion
                return
```

File: packages/web3-providers/src/zora/index.ts [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+
+    async getHistory(address: string, tokenId: string): Promise<NonFungibleTokenAPI.History[]> {
+        const variables = {
+            address,
+            tokenId,
+        }
+
+        const nftEventHistory = await this.client.request(getTokenHistoryQuery, variables)
+
+        const history = nftEventHistory.Token[0].transferEvents.map(
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/index.ts]:
```suggestion
        const history: NonFungibleTokenAPI.History[] = nftEventHistory.Token[0].transferEvents.map(
```

File: packages/web3-providers/src/zora/index.ts [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+                                link: `https://zora.co/${mint?.address}`,
+                            },
+                            to: {
+                                username: '',
+                                address: mint?.creator,
+                                imageUrl: '',
+                                link: `https://zora.co/${mint?.creator}`,
+                            },
+                        },
+                    } as NonFungibleTokenAPI.History
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/index.ts]: Try not to force typecasting.

File: packages/web3-providers/src/zora/index.ts [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+                    order_hash: ask?.transactionHash || '',
+                    maker_account: {
+                        user: { username: ask?.tokenOwner || '' },
+                        address: ask?.tokenOwner || '',
+                        profile_img_url: '',
+                        link: `https://zora.co/${ask?.tokenOwner}`,
+                    },
+                }
+            }
+            return {} as NonFungibleTokenAPI.AssetOrder
```
Comment (guanbinrui) [inline packages/web3-providers/src/zora/index.ts]:
```suggestion
            return
```

**Review by guanbinrui — DISMISSED** [review]

**Comment by brymut** [comment]
@Jack-Works could you please have a look at this?

**Comment by brymut** [comment]
@guanbinrui could you please re-approve, had to fix conflict with new dev branch changes

**Review by guanbinrui — APPROVED** [review]

**Review by Jack-Works — APPROVED** [review]

**Comment by brymut** [comment]
Hi @guanbinrui is this ready to merge?

**Comment by guanbinrui** [comment]
@DimensionDev/qa

**Comment by orange187** [comment]
@brymut When Collectibles switches to Zora, no data can be obtained and an error message is displayed （Failed to load your collectible on Zora.）

**Comment by brymut** [comment]
> @brymut When Collectibles switches to Zora, no data can be obtained and an error message is displayed （Failed to load your collectible on Zora.）

Could you please provide example link? @orange187

**Comment by brymut** [comment]
For example, https://twitter.com/damedoteth/status/1487966752579403776 this tweet works @orange187

https://user-images.githubusercontent.com/14836056/158790035-af64fb8b-dfe2-4fb3-9912-42cf41e5c82e.mov

**Comment by orange187** [comment]
@brymut   https://twitter.com/AkinoMush1/status/149975710155121869

https://user-images.githubusercontent.com/95333758/158928430-88c17839-015b-4524-b973-fb53b011ec23.mov

**Comment by brymut** [comment]
> @brymut   https://twitter.com/AkinoMush1/status/149975710155121869
> 
> 
> 
> https://user-images.githubusercontent.com/95333758/158928430-88c17839-015b-4524-b973-fb53b011ec23.mov
> 
> 
@orange187 that tweet has been deleted, is there another one? image

**Comment by orange187** [comment]
@brymut   Using the test connection you provided, the image does not load properly either

https://user-images.githubusercontent.com/95333758/158929350-a5517b6d-b86f-4e83-870b-55a7fb0e60b3.mov

**Comment by orange187** [comment]
@brymut https://twitter.com/AkinoMush1/status/1499757101551218691

**Comment by brymut** [comment]
> @brymut https://twitter.com/AkinoMush1/status/1499757101551218691

This is an opensea link, not Zora

**Comment by brymut** [comment]
> @brymut   Using the test connection you provided, the image does not load properly either
> 
> 
> 
> https://user-images.githubusercontent.com/95333758/158929350-a5517b6d-b86f-4e83-870b-55a7fb0e60b3.mov
> 
> 
> 
> 

Error in logs of your screen recording not coming from Collectible plug-in and I can't recreate the same error.

**Comment by orange187** [comment]
@brymut  But the pictures on my side can't load normally https://twitter.com/AkinoMush1/status/1504659688717447168
image

**Comment by brymut** [comment]
> @brymut But the pictures on my side can't load normally https://twitter.com/AkinoMush1/status/1504659688717447168 image

Works for me

https://user-images.githubusercontent.com/14836056/158933268-db947f23-bc5a-4f89-98ea-7fa61577199a.mov

Also, I've noticed that the logo on the selector is not rendered properly in your screenshots, that was an old error I had that I have already fixed, please check that you are building up to date pr-branch, that might be the problem @orange187

**Comment by orange187** [comment]
> 

Ok, let me see

**Comment by brymut** [comment]
Hi @AinuX could you please have a look too

**Comment by orange187** [comment]
@brymut  Switch to Zora, the bottom is not shown make an offer
image
image

**Comment by brymut** [comment]
> @brymut  Switch to Zora, the bottom is not shown make an offer
> 
> image
> 
> image
> 
> 

Make offer functionality only available on Opensea, Zora's api doesn't not have endpoint. Just like how Rarible doesn't have either. @orange187

**Comment by orange187** [comment]
> > @brymut  Switch to Zora, the bottom is not shown make an offer
> > image
> > image
> 
> Make offer functionality only available on Opensea, Zora's api doesn't not have endpoint. Just like how Rarible doesn't have either. @orange187

Oh, sorry to see the requirement just now. Yes, so far only NFT display and jump like are correct

**Comment by orange187** [comment]
Checked

**Pull request merged**